src/Analysis/Ast/Impl/Types/PythonClassType.cs

```diff
@@ -104,19 +105,35 @@ internal PythonClassType(string name, Location location)
 
         public override string Documentation {
             get {
-                // Try doc from the type (class definition AST node).
-                var doc = base.Documentation;
-                // Try docs __init__.
-                if (string.IsNullOrEmpty(doc)) {
-                    var init = GetMember("__init__") as IPythonFunctionType;
-                    doc = init?.DeclaringType == this ? init.Documentation : null;
+                if (!string.IsNullOrEmpty(_documentation)) {
+                    return _documentation;
                 }
-                // Try bases.
-                if (string.IsNullOrEmpty(doc) && Bases != null) {
-                    var o = DeclaringModule.Interpreter.GetBuiltinType(BuiltinTypeId.Object);
-                    doc = Bases.Except(Enumerable.Repeat(o, 1)).FirstOrDefault(b => !string.IsNullOrEmpty(b?.Documentation))?.Documentation;
+                // Make sure we do not cycle through bases back here.
+                if (!Push(this)) {
```

Comment:
Is it possible for this set to be in use for something else when the documentation is being calculated? Should there be a second one specifically for documentation accesses?

Reply:
Push/Pop should be under lock

Reply:
FWIW my comment here wasn't to do with locking, it was if Documentation was accessed during another call that was using Push/Pop to prevent infinite recursion where some class is visited, then somewhere down the callstack its documentation accessed (which will fail, because now Push returns false).